probes/requiresPRsToChangeCode/impl.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 OpenSSF Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//nolint:stylecheck
+package requiresPRsToChangeCode
+
+import (
+	"embed"
+	"errors"
+	"fmt"
+
+	"github.com/ossf/scorecard/v4/checker"
+	"github.com/ossf/scorecard/v4/finding"
+	"github.com/ossf/scorecard/v4/probes/internal/utils/uerror"
+)
+
+//go:embed *.yml
+var fs embed.FS
+
+const Probe = "requiresPRsToChangeCode"
+
+var errWrongValue = errors.New("wrong value, should not happen")
+
+func Run(raw *checker.RawResults) ([]finding.Finding, string, error) {
+	if raw == nil {
+		return nil, "", fmt.Errorf("%w: raw", uerror.ErrNil)
+	}
+
+	r := raw.BranchProtectionResults
+	var findings []finding.Finding
+
+	if len(r.Branches) == 0 {
+		f, err := finding.NewWith(fs, Probe, "no branches found", nil, finding.OutcomeNotApplicable)
+		if err != nil {
+			return nil, Probe, fmt.Errorf("create finding: %w", err)
+		}
+		findings = append(findings, *f)
+		return findings, Probe, nil
+	}
+
+	for i := range r.Branches {
+		branch := &r.Branches[i]
+
+		protected := !(branch.Protected != nil && !*branch.Protected)
+		var protectedValue int
+		if protected {
+			protectedValue = 1
+		} else {
+			protectedValue = 0
+		}
+
+		nilMsg := fmt.Sprintf("could not determine whether branch '%s' requires PRs to change code", *branch.Name)
+		trueMsg := fmt.Sprintf("PRs are required in order to make changes on branch '%s'", *branch.Name)
+		falseMsg := fmt.Sprintf("PRs are not required to make changes on branch '%s'; ", *branch.Name) +
+			"or we don't have data to detect it." +
+			"If you think it might be the latter, make sure to run Scorecard with a PAT or use Repo " +
+			"Rules (that are always public) instead of Branch Protection settings"
```

Comment:
@diogoteles08 how attached are you to the `false` reason string here mentioning we may be unable to detect it? Otherwise we could probably simplify by using the `GetTextOutcomeFromBool` helper we use elsewhere. Which would just say "PRs are not required" (or something similar)

Reply:
I personally care about this clarification, because I AFAIK this is the only case where we give a negative response when the project might even be doing everything right (for other cases we make it clear we don't have data and we don't penalize them). The documentation doesn't make it clear either, so that can bring frustration to maintainers as they might not understand what they're doing wrong.

Maybe a alternative would be to review [Branch protection documentation](https://github.com/ossf/scorecard/blob/main/docs/checks.md#branch-protection) to make that explicit. I.e., at the line we mention the rule of Require PRs, we add a note saying something like "when using GitHub traditional Branch Protections without admin permissions, we can't differentiate the case of zero reviewer from not requiring PRs. That's due to limitations on GitHub API and causes a false negative when requiring PRs without requiring reviewers."